AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Endpoint/EndpointClient.swift

```diff
@@ -285,6 +285,42 @@ actor EndpointClient: EndpointClientBehaviour {
             }
         }
     }
+    
+    private static func getStoredData(
+        from keychain: KeychainStoreBehavior,
+        forKey key: String,
+        fallbackTo defaultSource: UserDefaultsBehaviour
+    ) -> Data? {
+        if let data = try? keychain.getData(key) {
+            return data
+        } else {
+            return defaultSource.data(forKey: key)
+        }
+    }
+    
+    private static func getStoredAttributes(
```

Comment:
nit: `getStoredAttributes` and `getStoredMetrics` are pretty much the same thing, just with different keys and return types. 

You can use a generic version instead:

```
private static func get<T: Decodable>(
    key: String,
    as: T.Type,
    from keychain: KeychainStoreBehavior,
    fallbackTo defaultSource: UserDefaultsBehaviour,
    using archiver: AmplifyArchiverBehaviour
) -> T? {
    guard let data = try? keychain.getData(key) else {
        return defaultSource.object(forKey: key) as? T
    }

    return try? archiver.decode(T.self, from: data)
}
```
---
Also, if you don't wanna be passing around 3 storage arguments each time on these retrieval functions, you can follow what we did on `PinpointContext` [and have a `struct` that groups all of them](https://github.com/aws-amplify/amplify-ios/blob/dev-preview/AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/PinpointContext.swift#L111-L116), so you can just [store that as a property](https://github.com/aws-amplify/amplify-ios/blob/dev-preview/AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/PinpointContext.swift#L134-L137) and pass it around like 
```
let token = Self.getData(forKey: Constants.deviceTokenKey, from: storage)
``` 

But these are suggestions, up to you :)

Reply:
I combined the two functions to be generic and leaving the arguments as separate for now to keep to keep the argument names clear instead of combining them into a re-usable struct.